File: php/src/misskey.php
<?php

declare(strict_types=1);

function misskey_content_type(string $mime): string { return str_starts_with($mime, 'image/') ? 'image' : (str_starts_with($mime, 'audio/') ? 'audio' : (str_starts_with($mime, 'video/') ? 'video' : 'file')); }
/** @return array<string,mixed> */
function misskey_parse_event(array $payload): array
{
    $note = $payload['note'] ?? $payload; $userId = $note['userId'] ?? $note['user']['id'] ?? ''; if ($userId === '') throw new InvalidArgumentException('Misskey note has no user'); $file = $note['files'][0] ?? null; return ['platform' => 'misskey', 'user_id' => (string)$userId, 'content_type' => $file ? misskey_content_type((string)($file['type'] ?? '')) : 'text', 'text' => $note['text'] ?? null, 'media_url' => $file['id'] ?? null, 'reply_to_id' => isset($note['id']) ? (string)$note['id'] : null];
}
function misskey_send_reply(array $event, array $reply, string $baseUrl, string $token): void
{
    if ($baseUrl === '' || $token === '') throw new InvalidArgumentException('Misskey base URL and token are required'); foreach (array_slice($reply['messages'] ?? [], 0, 5) as $message) { $body = ['i' => $token, 'text' => $message['text'] ?? '', 'replyId' => $event['reply_to_id'] ?? null]; if (($message['type'] ?? 'text') !== 'text') { if (empty($message['media_url'])) throw new InvalidArgumentException('Misskey media reply requires file ID'); $body['fileIds'] = [$message['media_url']]; } $ch = curl_init(rtrim($baseUrl, '/').'/api/notes/create'); curl_setopt_array($ch, [CURLOPT_POST => true, CURLOPT_RETURNTRANSFER => true, CURLOPT_HTTPHEADER => ['Content-Type: application/json'], CURLOPT_POSTFIELDS => json_encode($body, JSON_THROW_ON_ERROR)]); $result = curl_exec($ch); $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE); if ($result === false || $status < 200 || $status >= 300) throw new RuntimeException('Misskey API request failed'); curl_close($ch); }
}

File: python/platforms/misskey.py
from __future__ import annotations

import os
from typing import Any

import requests

from core.models import InboundEvent, OutboundReply
from .base import PlatformAdapter
from .catalog import PLATFORM_CATALOG


class MisskeyAdapter(PlatformAdapter):
    """Misskey HTTP API adapter for notes and note media links."""

    capabilities = PLATFORM_CATALOG["misskey"]

    def __init__(self, base_url: str | None = None, token: str | None = None, session: requests.Session | None = None):
        self.base_url = (base_url or os.getenv("MISSKEY_BASE_URL", "")).rstrip("/")
        self.token = token or os.getenv("MISSKEY_TOKEN")
        if not self.base_url or not self.token:
            raise ValueError("MISSKEY_BASE_URL and MISSKEY_TOKEN are required")
        self.session = session or requests.Session()

    def parse_event(self, payload: Any, headers: dict[str, str] | None = None) -> InboundEvent:
        note = payload.get("body", payload.get("note", payload))
        user_id = str(note.get("userId") or note.get("user", {}).get("id", ""))
        if not user_id:
            raise ValueError("Misskey note has no user")
        files = note.get("files") or []
        common = {"platform": "misskey", "user_id": user_id, "reply_to_id": str(note.get("id")) if note.get("id") else None}
        if note.get("text"):
            return InboundEvent(**common, content_type="text", text=note["text"])
        if files:
            file = files[0]
            content_type = self._content_type(file.get("type", ""))
            return InboundEvent(**common, content_type=content_type, media_url=file.get("url"))
        raise ValueError("Misskey note has no supported content")

    def send_reply(self, event: InboundEvent, reply: OutboundReply) -> None:
        for message in reply.messages[: self.capabilities.max_reply_items or len(reply.messages)]:
            text = message.text or message.media_url or ""
            response = self.session.post(
                f"{self.base_url}/api/notes/create",
                json={"i": self.token, "text": text[:3000], **({"replyId": event.reply_to_id} if event.reply_to_id else {})},
                timeout=30,
            )
            response.raise_for_status()
            body = response.json()
            if body.get("error"):
                raise RuntimeError(f"Misskey API error: {body['error']}")

    @staticmethod
    def _content_type(mime_type: str) -> str:
        if mime_type.startswith("image/"):
            return "image"
        if mime_type.startswith("audio/"):
            return "audio"
        if mime_type.startswith("video/"):
            return "video"
        return "file"
